shared/eventbus/eventbus.go

```diff
@@ -21,34 +26,120 @@ type EventType struct {
+	id      string
+	handler Handler
+}
+
 type BrokerService struct {
-	QueueLimit     int
-	GoroutineLimit int
+	queueLimit      int
+	goroutineLimit  int
+	subscribers     map[string][]subscriber
+	subscriberMutex *sync.Mutex
+	channel         chan Event
+	eventTypes      map[string]EventType
+	eventMutex      *sync.Mutex
 }
 
 func NewBroker(queueLimit, goroutineLimit int) *BrokerService {
 	return &BrokerService{
-		QueueLimit:     queueLimit,
-		GoroutineLimit: goroutineLimit,
+		queueLimit:      queueLimit,
+		goroutineLimit:  goroutineLimit,
+		subscribers:     map[string][]subscriber{},
+		subscriberMutex: &sync.Mutex{},
+		channel:         make(chan Event, queueLimit),
+		eventMutex:      &sync.Mutex{},
 	}
 }
 
 func (b *BrokerService) Register(topic, description string, typ any) error {
+	schema := jsonschema.Reflect(typ)
+	buf, err := schema.MarshalJSON()
+	if err != nil {
+		return errors.Wrap(err, "can't register the topic")
+	}
+
+	evT := EventType{
+		Topic:       topic,
+		Description: description,
+		Schema:      string(buf),
+	}
+
+	b.eventMutex.Lock()
+	defer b.eventMutex.Unlock()
+	b.eventTypes[topic] = evT
 	return nil
 }
-func (b *BrokerService) EventTypes() ([]EventType, error) {
-	return nil, nil
+
+func (b *BrokerService) EventTypes() []EventType {
+	types := make([]EventType, 0, len(b.eventTypes))
+	for _, event := range b.eventTypes {
+		types = append(types, event)
+	}
+	return types
 }
 
 func (b *BrokerService) Publish(topic string, ctx request.CTX, data any) error {
+	if _, ok := b.eventTypes[topic]; !ok {
+		return errors.New("topic does not exist")
+	}
+	ev := Event{
+		Topic:   topic,
+		Context: ctx,
+		Message: data,
+	}
+
+	// run async not to block a caller
+	go func() {
+		b.channel <- ev
+	}()
 	return nil
 }
+
 func (b *BrokerService) Subscribe(topic string, handler Handler) (string, error) {
-	return "", nil
+	b.subscriberMutex.Lock()
+	defer b.subscriberMutex.Unlock()
+
+	if _, ok := b.eventTypes[topic]; !ok {
+		return "", errors.New("topic does not exist")
+	}
+
+	id := model.NewId()
+	b.subscribers[topic] = append(b.subscribers[topic], subscriber{
+		id:      id,
+		handler: handler,
+	})
+	return id, nil
 }
+
 func (b *BrokerService) Unsubscribe(topic, id string) error {
-	return nil
+	b.subscriberMutex.Lock()
+	defer b.subscriberMutex.Unlock()
+
+	if _, ok := b.eventTypes[topic]; !ok {
+		return errors.New("topic does not exist")
+	}
+
+	subscribers := b.subscribers[topic]
+	for i, subscriber := range subscribers {
+		if subscriber.id == id {
+			newSubscribers := append(subscribers[:i], subscribers[i+1:]...)
+			b.subscribers[topic] = newSubscribers
+			return nil
+		}
+	}
+	return errors.New("id was not found")
 }
-func (b *BrokerService) Start() error {
-	return nil
+
+func (b *BrokerService) Start() {
+	go b.runHandlers()
+}
+
+func (b *BrokerService) runHandlers() {
+	for {
+		ev := <-b.channel
+		for _, subscriber := range b.subscribers[ev.Topic] {
+			go subscriber.handler(ev)
+		}
+	}
```

Comment:
This will never exit from the loop, we can at least make it as: following:
```suggestion
	for ev := range b.channel {
		for _, subscriber := range b.subscribers[ev.Topic] {
			go subscriber.handler(ev)
		}
	}
```

or we can even use select if we add a timeout or cancellations signal like:
```Go
for {
	select {
	case ev := <-b.channel:
		...
	case <-b.cancel:
		fmt.Println("quiting")
		return
	}
}
```

Reply:
We will address this in a separate PR.